quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,681 @@
+
+/// The duration that the pulsar consumer waits for a new message
+/// to be dispatched.
+const MESSAGE_WAIT_TIMEOUT: Duration = Duration::from_secs(1);
+
+type ConsumerMessage = Result<Message<PulsarMessage>, pulsar::Error>;
+
+pub struct PulsarSourceFactory;
+
+#[async_trait]
+impl TypedSourceFactory for PulsarSourceFactory {
+    type Source = PulsarSource;
+    type Params = PulsarSourceParams;
+
+    async fn typed_create_source(
+        ctx: Arc<SourceExecutionContext>,
+        params: PulsarSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self::Source> {
+        PulsarSource::try_new(ctx, params, checkpoint).await
+    }
+}
+
+#[derive(Default)]
+pub struct PulsarSourceState {
+    /// Number of bytes processed by the source.
+    pub num_bytes_processed: u64,
+    /// Number of messages processed by the source (including invalid messages).
+    pub num_messages_processed: u64,
+    /// Number of invalid messages, i.e., that were empty or could not be parsed.
+    pub num_invalid_messages: u64,
+}
+
+pub struct PulsarSource {
+    ctx: Arc<SourceExecutionContext>,
+    pulsar: PulsarConsumer,
+    params: PulsarSourceParams,
+    state: PulsarSourceState,
+
+    partition_id: PartitionId,
+    previous_position: Position,
+}
+
+impl PulsarSource {
+    pub async fn try_new(
+        ctx: Arc<SourceExecutionContext>,
+        params: PulsarSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let mut builder = Pulsar::builder(&params.address, TokioExecutor);
+
+        match params.authentication.clone() {
+            None => {}
+            Some(PulsarSourceAuth::Token(token)) => {
+                let auth = Authentication {
+                    name: "token".to_string(),
+                    data: token.as_bytes().to_vec(),
+                };
+
+                builder = builder.with_auth(auth);
+            }
+            Some(PulsarSourceAuth::Oauth2 {
+                issuer_url,
+                credentials_url,
+                audience,
+                scope,
+            }) => {
+                let auth = OAuth2Params {
+                    issuer_url,
+                    credentials_url,
+                    audience,
+                    scope,
+                };
+                builder =
+                    builder.with_auth_provider(OAuth2Authentication::client_credentials(auth));
+            }
+        }
+
+        info!(
+            index_id=%ctx.index_id,
+            source_id=%ctx.source_config.source_id,
+            topic=%params.topic,
+            "Starting Pulsar source."
+        );
+
+        let pulsar: Pulsar<_> = builder.build().await?;
+
+        let partition_id = PartitionId::from(params.topic.as_str());
+        let seek_to = checkpoint
+            .position_for_partition(&partition_id)
+            .and_then(msg_id_from_position);
+        let pulsar = spawn_message_listener(params.clone(), pulsar, seek_to).await?;
```

Comment:
It's worth noting here that we're using the topic name as the partition ID currently as there are some outstanding questions I'm still looking into around how Pulsar behaves when seeking and acking messages in partitions.

From what I understand it is not really designed to be used in situations where you're using seeking like we are to do the at least once semantics, and because of that, I am still trying to understand how to go about handling partitioned topics without explicitly declaring each indexer's partition and without causing us to accidentally loose messages.